library/ssl_tls12_server.c

```diff
@@ -2808,6 +2871,45 @@ static int ssl_prepare_server_key_exchange( mbedtls_ssl_context *ssl,
     if( ciphersuite_info->key_exchange == MBEDTLS_KEY_EXCHANGE_ECJPAKE )
     {
         int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+        unsigned char *out_p = ssl->out_msg + ssl->out_msglen;
+        unsigned char *end_p = ssl->out_msg + MBEDTLS_SSL_OUT_CONTENT_LEN -
+                               ssl->out_msglen;
+        size_t output_offset = 0;
+        size_t output_len = 0;
+        const mbedtls_ecp_curve_info *curve_info;
+
+        /*
+         * The first 3 bytes are:
```

Comment:
I'll note an asymmetry here: for parsing those 3 bytes are handled by the round function, while for writing they aren't. I think it would be better to have symmetry. I have a slight preference for handling them out of the `round` function (as it done here) as it saves the trouble of dynamically checking whether we're client or server: we know based on whether we're in `ssl_tls12_client.c` or `ssl_tls12_server.c`. Also, it saves a bit of code in builds that have only client or only server, for the same reason.

Reply:
I just realized that after this change the two read functions for round 1 and 2 are actually identical so I'm planning to shrink them together. The same also for the writing one, of course